### PR TITLE
subsys/settings: Allow to register custom backends in settings subsystem

### PR DESCRIPTION
**Pull request by fnde-ot**
This is a rework of the settings subsystem based on closed PR #12998 

Major changes are:
- Expose settings backend API to enable custom backend support.
- Add a new CONFIG_SETTINGS_CUSTOM backend to allow registering a custom
backend.
- Change api of the handlers h_set() routines to allow for
backend-specific read callbacks.
- Provide a customizable settings_backend_init() routine for custom
backends.
- Move runtime settings support to be its own backend.

Signed-off-by: François Delawarde (FNDE) <fnde@oticon.com>

### CODE REVIEW

**Review by Vudentz — CHANGES_REQUESTED** [review]
It looks like there are a lot of unrelated changes to bluetooth, mostly code style changes which probably should be either left alone or submited in a different patch.

**Comment by fnde-ot** [comment]
> It looks like there are a lot of unrelated changes to bluetooth, mostly code style changes which probably should be either left alone or submited in a different patch.

@Vudentz, Sorry about that, I followed the contribution guidelines and ran uncrustify and checkpatch on all files in the diff, and the previous contributor might have skipped that step. I will try to move those unrelated code style changes into another commit.

File: doc/reference/storage/settings/settings.rst [hunk 1]
```diff
@@ -40,11 +40,37 @@ These are registered using a call to ``settings_register()``.
...
+
+**csi_save_end**
+    This gets called after having saved of all current settings using
+    ``settings_save()``.
+
+Zephyr Storage Backends
+***********************
+
+Zephyr has two existing backend storages which can be a Flash Circular Buffer
+(CONFIG_SETTINGS_FCB) or a file in the filesystem (CONFIG_SETTINGS_FS).
```
Comment (dbkinder) [inline doc/reference/storage/settings/settings.rst]: Link to the option documentation:
```suggestion
(:option:`CONFIG_SETTINGS_FCB`) or a file in the filesystem (:option:`CONFIG_SETTINGS_FS`).
```

File: doc/reference/storage/settings/settings.rst [hunk 2]
```diff
@@ -152,6 +178,37 @@ up from where it was before restart.
         sys_reboot(SYS_REBOOT_COLD);
     }
 
+Example: Custom Backend Implementation
+**************************************
+
+This is a simple example showing how to register a simple custom backend
+handler (CONFIG_SETTINGS_CUSTOM).
```
Comment (dbkinder) [inline doc/reference/storage/settings/settings.rst]:
```suggestion
handler (:option:`CONFIG_SETTINGS_CUSTOM`).
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
Looks good with a couple of edits

File: doc/reference/storage/settings/settings.rst (hunk 1 above)
Comment (fnde-ot) [reply doc/reference/storage/settings/settings.rst]: Fixed, thanks!

File: doc/reference/storage/settings/settings.rst (hunk 2 above)
Comment (fnde-ot) [reply doc/reference/storage/settings/settings.rst]: Fixed, thanks!

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, thanks.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c [hunk 3]
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
...
+	size_t len = 0;
 
 	if (argc == 1) {
 		if (!strcmp(argv[0], "rc")) {
-			len = settings_val_read_cb(val_ctx, &reset_counter,
-						   sizeof(reset_counter));
+			len = read_cb(
+			 cb_arg,
+			 &reset_counter,
+			 sizeof(reset_counter));
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c]: This looks like very strange style - surely you can fit at least some of these on the same line?

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c [hunk 4]
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
...
 
 		if (!strcmp(argv[0], "gdtt")) {
-			len = settings_val_read_cb(
-				val_ctx,
-				&gen_def_trans_time_srv_user_data.tt,
-				sizeof(gen_def_trans_time_srv_user_data.tt));
+			len = read_cb(
+			 cb_arg,
+			 &gen_def_trans_time_srv_user_data.tt,
+			 sizeof(gen_def_trans_time_srv_user_data.tt));
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c]: Same here, though I realise the old code in this case had the strange style as well.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c [hunk 5]
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
...
 
 		if (!strcmp(argv[0], "gpo")) {
-			len = settings_val_read_cb(
-			val_ctx,
-			&gen_power_onoff_srv_user_data.onpowerup,
-			sizeof(gen_power_onoff_srv_user_data.onpowerup));
+			len = read_cb(
+			 cb_arg,
+			 &gen_power_onoff_srv_user_data.onpowerup,
+			 sizeof(gen_power_onoff_srv_user_data.onpowerup));
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c]: And here - put arguments which fit on the same line on the same line

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c [hunk 6]
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
...
 
 		if (!strcmp(argv[0], "ltd")) {
-			len = settings_val_read_cb(
-			val_ctx,
-			&light_ctl_srv_user_data.lightness_temp_def,
-			sizeof(light_ctl_srv_user_data.lightness_temp_def));
+			len = read_cb(
+			 cb_arg,
+			 &light_ctl_srv_user_data.lightness_temp_def,
+			 sizeof(light_ctl_srv_user_data.lightness_temp_def));
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c]: and here

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c [hunk 7]
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
...
 
 		if (!strcmp(argv[0], "ltl")) {
-			len = settings_val_read_cb(
-			val_ctx,
-			&light_ctl_srv_user_data.lightness_temp_last,
-			sizeof(light_ctl_srv_user_data.lightness_temp_last));
+			len = read_cb(
+			 cb_arg,
+			 &light_ctl_srv_user_data.lightness_temp_last,
+			 sizeof(light_ctl_srv_user_data.lightness_temp_last));
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c]: and here

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c [hunk 8]
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
...
 
 		if (!strcmp(argv[0], "lr")) {
-			len = settings_val_read_cb(
-			val_ctx,
-			&light_lightness_srv_user_data.lightness_range,
-			sizeof(light_lightness_srv_user_data.lightness_range));
+			len = read_cb(
+			 cb_arg,
+			 &light_lightness_srv_user_data.lightness_range,
+			 sizeof(light_lightness_srv_user_data.lightness_range));
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c]: and here

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c [hunk 9]
```diff
@@ -114,56 +114,59 @@ void save_on_flash(u8_t id)
...
 
 		if (!strcmp(argv[0], "tr")) {
-			len = settings_val_read_cb(
-			val_ctx,
-			&light_ctl_srv_user_data.temperature_range,
-			sizeof(light_ctl_srv_user_data.temperature_range));
+			len = read_cb(
+			 cb_arg,
+			 &light_ctl_srv_user_data.temperature_range,
+			 sizeof(light_ctl_srv_user_data.temperature_range));
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/storage.c]: and here

**Review by jhedberg — CHANGES_REQUESTED** [review]
Looks generally fine to me, except for some coding style issues in the onoff_level_lightning_vnd_app

**Review by jhedberg — APPROVED** [review]

**Comment by carlescufi** [comment]
@nvlsianpu can you take a look at this one?

**Review by Vudentz — APPROVED** [review]

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: doc/reference/storage/settings/settings.rst [hunk 10]
```diff
@@ -40,11 +40,38 @@ These are registered using a call to ``settings_register()``.
     when ``settings_save()`` tries to save the settings or transfer to any
     user-implemented back-end.
 
-Persistence
-***********
+Backends
+********
+
+Backends are meant to load and save data from/to persistent storage, and
```
[the remark this answers is not in the document]
Comment (fnde-ot) [reply doc/reference/storage/settings/settings.rst]: True that!

File: include/settings/settings.h [hunk 11]
```diff
@@ -151,88 +153,58 @@ int settings_save_one(const char *name, void *value, size_t val_len);
...
- * Config storage
+ * API for config storage
  */
 struct settings_store_itf;
 struct settings_store {
 	sys_snode_t cs_next;
 	const struct settings_store_itf *cs_itf;
 };
 
+struct settings_store_itf {
```
[the remark this answers is not in the document]
Comment (fnde-ot) [reply include/settings/settings.h]: Will do.

File: include/settings/settings.h [hunk 12]
```diff
@@ -151,88 +153,58 @@ int settings_save_one(const char *name, void *value, size_t val_len);
...
+	int (*csi_load)(struct settings_store *cs);
```
[the remark this answers is not in the document]
Comment (fnde-ot) [reply include/settings/settings.h]: I agree, so do the export and get functions in module handlers, but I would rather that being done in another PR if it is okay.

[the remark this answers is not in the document]
Comment (nvlsianpu) [reply include/settings/settings.h]: Ok, maybe we can do that in coming PRs (will see)

File: subsys/settings/Kconfig [hunk 13]
```diff
@@ -35,7 +38,7 @@ config SETTINGS_USE_BASE64
 
 choice
 	prompt "Storage back-end"
-	default SETTINGS_FCB if FCB
+	default SETTINGS_NONE
```
[the remark this answers is not in the document]
Comment (fnde-ot) [reply subsys/settings/Kconfig]: It makes a lot of qemu sanitycheck tests fail. For example samples/bluetooth/peripheral which selects FCB even though it is not supported for qemu_x86. That change makes it fall back to the default SETTINGS_NONE.

[the remark this answers is not in the document]
Comment (fnde-ot) [reply subsys/settings/Kconfig]: Resolved by putting both lines in, so it falls back to SETTINGS_NONE if there FCB is not enabled.

**Review by nvlsianpu — APPROVED** [review]

**Comment by jukkar** [comment]
Merging this one as requested by @aescolar as most of the needed people are in wrong timezone.

**Pull request merged**